### PR TITLE
Feature Add Metadata To Report Via File Pattern

### PR DESCRIPTION
**Pull request by bhreinb**

### CODE REVIEW

**Comment by damianszczepanik** [comment]
I would be happy to have this feature in core of the report https://github.com/damianszczepanik/cucumber-reporting

**Comment by bhreinb** [comment]
I haven't looked into this module all that much to be honest with you but happy to have a look...but from having a quick look there seems to be 2 possibilities as I see it

`1) Add new method in configuration class something like so
configuration.addClassificationsFromMetaDataFilePattern
and this would populate our list of classifications?
`
or

`2) pass in a list of properties files to the ReportBuilder class
the constructor of the ReportBuilder class would change from two parameters to three and ReportParser class would get a new method called parsePropertiesFile to add classifications got from properties file`

let me know your preference

**Comment by damianszczepanik** [comment]
I would like to accept list of property files so as the developer I don't care about extracting values from the files so only what I need is to build proper list with properties. Similar to https://github.com/damianszczepanik/cucumber-reporting/blob/master/src/main/java/net/masterthought/cucumber/ReportBuilder.java#L65

Does it make sense?

**Comment by bhreinb** [comment]
Yes that makes sense & aligns with 2) above. I'll try to get to this as soon as I can. Essentially it's 2 pull requests one for each of the two repos (cumcumber reporting & cucumber reporting Jenkins).

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 1]
```diff
@@ -38,6 +38,7 @@
 public class CucumberReportPublisher extends Publisher implements SimpleBuildStep {
 
     private final static String DEFAULT_FILE_INCLUDE_PATTERN = "**/*.json";
+    private final static String DEFAULT_FILE_INCLUDE_PATTERN_PROPERTIES = "**/*.properties";
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: `_CLASSIFICATIONS`

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 2]
```diff
@@ -212,6 +213,15 @@ public String getSortingMethod() {
         return sortingMethod;
     }
 
+    @DataBoundSetter
+    public void setMetaDataFilePattern(String metaDataFilePattern) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: same

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 3]
```diff
@@ -212,6 +213,15 @@ public String getSortingMethod() {
...
+        this.metaDataFilePattern = metaDataFilePattern;
+    }
+
+    public String getMetaDataFilePattern() {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: same

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 4]
```diff
@@ -271,14 +286,15 @@ private void generateReport(Run<?, ?> build, FilePath workspace, TaskListener li
...
-        }
+        log(listener, String.format("Adding %d classifications sourced from jenkins configuration interface!", classifications.size()));
+
+        addClassificationsToBuildReport(build,workspace,listener,configuration,classifications);
+
+        List<Classification> sourcedFromMetaDataFiles = fetchMetaDataClassifications(directoryJsonCache, listener);
+
+        log(listener, String.format("Adding %d classifications sourced from metadata file pattern '%s'!",sourcedFromMetaDataFiles.size(),metaDataFilePattern));
+
+        addClassificationsToBuildReport(build,workspace,listener,configuration,sourcedFromMetaDataFiles);
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: spaces after commas

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 5]
```diff
@@ -369,6 +385,98 @@ private String evaluateMacro(Run<?, ?> build, FilePath workspace, TaskListener l
...
+        }
+
+    }
+
+    private List<Classification> fetchMetaDataClassifications(File targetDirectory, TaskListener listener) throws IOException {
+
+        List<Classification> allSourcedFromFiles = new ArrayList<>();
+
+        DirectoryScanner scanner = new DirectoryScanner();
+
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: group lines, remove useless empty lines

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 6]
```diff
@@ -369,6 +385,98 @@ private String evaluateMacro(Run<?, ?> build, FilePath workspace, TaskListener l
         return value;
     }
 
+    private void addClassificationsToBuildReport(Run<?, ?> build, FilePath workspace, TaskListener listener, Configuration configuration, List<Classification> listToAdd) throws InterruptedException, IOException {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: make sure that `listToAdd` is not null earlier...

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 7]
```diff
@@ -369,6 +385,98 @@ private String evaluateMacro(Run<?, ?> build, FilePath workspace, TaskListener l
...
+
+        if (CollectionUtils.isNotEmpty(listToAdd)) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: ... so this line can be removed

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java (hunk 1 above)
Comment (bhreinb) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: btw do you want to keep the idea of having a default search pattern for classification files or should it be removed? I'm going with assumption keep for now & we can update later if need be

Comment (bhreinb) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: I'm going with the assumption that all properties files within target directory are copied to the cache directory like what happens with the json files...but only classifications get added when a pattern is specified and files are found...make sense?

Comment (damianszczepanik) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: The main reason of cache was to be able to check what JSON file was used in previous build so reporting and debugging issues are easier. I don't think this is critical for classification so it's up to you

**Comment by bhreinb** [comment]
Hey @damianszczepanik artifact 3.12.0 isn't in maven (https://mvnrepository.com/artifact/net.masterthought/cucumber-reporting) as of yet but I tested against my local setup

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java (hunk 1 above)
Comment (bhreinb) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: I use this string to copy property files to the cache directory & for nothing else. So I suppose lets leave as it is :)

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 8]
```diff
@@ -38,6 +38,7 @@
 public class CucumberReportPublisher extends Publisher implements SimpleBuildStep {
 
     private final static String DEFAULT_FILE_INCLUDE_PATTERN = "**/*.json";
+    private final static String DEFAULT_FILE_INCLUDE_PATTERN_CLASSIFICATIONS = "**/*.properties";
```
[the remark this answers is not in the document]
Comment (bhreinb) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: This is still compatible with older versions as far as I know...all this string is used for is copying files that have ext type 'properties' to the cache location...then we match the file pattern against the cached location...do you still want it to be empty?

**Comment by damianszczepanik** [comment]
What happen after upgrade if you have some properties files in workspace? Now I expect they will be scanned and added to classifications. And this is not what we expect because after upgrade report should stay the same. Am I correct?

**Comment by bhreinb** [comment]
They need to have a valid setting for classificationsFilePattern for any classifications to be added. Default for this is empty.

`private String classificationsFilePattern = "";`

this only gets updated when the user intentionally adds a setting for 'classificationsFilePattern' via the jenkins interface.

to add any classifications via properties we have to invoke method

`fetchPropertyFiles`

we guard this for empty or null strings 👍 

`Line 397 if (StringUtils.isNotEmpty(classificationsFilePattern)) {`

sounds good?

**Comment by damianszczepanik** [comment]
Sounds good. Push PR and I will merge it when problem with azure is fixed.

**Comment by bhreinb** [comment]
you have the up to date PR :)...

**Comment by damianszczepanik** [comment]
@bhreinb change is merged, thanks! would you like add same support also for https://github.com/damianszczepanik/cucumber-sandwich and https://github.com/damianszczepanik/maven-cucumber-reporting ?

**Comment by bhreinb** [comment]
Hi @damianszczepanik no problem. Happy to support those other 2 projects but I won't get to them however for a few weeks or so...currently on holidays in Peru.

**Comment by damianszczepanik** [comment]
@bhreinb I'm looking forward to see support for this in both project so prepare PR and I will be happy to merge it

**Pull request merged**